### ISSUE TITLE
Error messages lead to stuck/unchanging pop-up in Inspect view

### ISSUE THREAD

**Issue by kateler** [maputnik/editor#609]
I just upgraded to the latest Maputnik (from 1.4, I think). Since the upgrade, I get a stream of errors in my browser console after enabling Inspect view and mousing over the map. Here's an example: 

```
react-dom.development.js:558 Warning: Encountered two children with the same key, `transportation_name-undefined`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
    in div (created by FeaturePropertyPopup)
    in FeaturePropertyPopup
warningWithoutStack	@	react-dom.development.js:558
```

I get a lot of these errors mentioning several of my map data layers. After mousing around briefly, the same handful of features is displayed in the Inspect pop-up no matter where I go on the map. 

Anyone know how to fix this? 

In case it's relevant, I also get this error once: 

```
Warning: Failed prop type: Invalid prop `value` of type `boolean` supplied to `StringInput`, expected `string`.
    in StringInput (created by FeaturePropertyPopup)
    in FeaturePropertyPopup
printWarning | @ | checkPropTypes.js:20
```

**Comment by pathmapper** [comment]
Hi @kateler,

which version of Maputnik are you using currently?

The latest release is `v1.6.1` (see https://github.com/maputnik/editor/releases) which is also running on https://maputnik.github.io/editor/

Could you please also check if you are seeing this error also with `master`? Here is a link to try it out:
https://2155-67726921-gh.circle-artifacts.com/0/artifacts/build/index.html

**Comment by kateler** [comment]
I cloned the repository earlier this week (commit 5b9af07ebc7864b660425f9cbd37b025bcb51d42), so it's newer than 1.6.1. Is that what you wanted me to try?

**Comment by pathmapper** [comment]
Thanks, I can't reproduce the issue with https://github.com/maputnik/editor/commit/5b9af07ebc7864b660425f9cbd37b025bcb51d42.

Could you please try if you can reproduce the issue when opening your style with 
https://2155-67726921-gh.circle-artifacts.com/0/artifacts/build/index.html (which was built with the latest commit https://github.com/maputnik/editor/commit/63ed8c1de34741432cc2adbfd9d1138756579186).

Also it would be helpful if you would provide your style where you are seeing this error.

**Comment by kateler** [comment]
Thanks, @pathmapper !

First, I have an update on the behavior I'm seeing. I originally described the Inspect pop-up as "stuck" because what I saw wasn't changing. I eventually noticed it had a scrollbar, and I think what's actually happening is that every feature I mouseover is being added to the pop-up. I can see the scrollbar size changing. After mousing around, I can scroll down and see hundreds (or more) features in the pop-up. 

When I tried the link you gave me above, the same behavior happened, but there were no errors in the browser console. The behavior was the same as I described in the previous paragraph. 

To test, I created an extremely simple stylesheet with no layers. I've pasted it below with just the tile URL obscured so you can see just how basic it is. With this stylesheet, I still get the same behavior on your link as well as my current Maputnik. When I run my previous Maputnik (1.4.0) and load the simple stylesheet, the issue doesn't happen. 

```
{
  "version": 8,
  "name": "Simplest Style",
  "metadata": {"amazon:note": "", "maputnik:renderer": "mbgljs"},
  "zoom": 4,
  "bearing": 0,
  "pitch": 0,
  "sources": {
    "Vector": {
      "type": "vector",
      "url": "https://url-goes-here.json"
    }
  },
  "layers": [
  ],
  "id": "simplest_style"
}

```

**Comment by pathmapper** [comment]
@kateler thanks for the update, however I can't still reproduce the behavior you are describing.

For me the inspect mode is working fine, please have a look at the following GIF where I tried it on the OpenMapTiles vector tile source:

![inspect_mode_hover](https://user-images.githubusercontent.com/20856381/73208145-391dec80-4146-11ea-92b8-21fae33ea040.gif)

Could you share a style where the tile URL is not obscured? To be able to diagnose the issue we need a reproducible test case which shows the issue.

**Comment by pathmapper** [comment]
Do you see the issue only with a certain vector tile source or does it happen also when you are using the inspect mode e.g. on the OSM Liberty style?

You could try it here:
https://maputnik.github.io/editor/?style=https://maputnik.github.io/osm-liberty/style.json

**Comment by kateler** [comment]
OK, good news. I can reproduce this using something you can try: 

1. Go to https://maputnik.github.io/editor/#9.85/51.4922/719.8773
2. Click Open
3. Click Zoomstack Outdoor
4. Change mode from Map to Inspect
5. Drag around the map slowly

What I notice is that at first, it works as it should. But eventually, certain roads get "stuck" at the top of the pop-up and remain there, and eventually the pop-up fills with those. 

I reproduced this in Chrome, Chrome incognito mode, and Firefox (same in each).

**Comment by orangemug** [comment]
@kateler thanks for the test case, I can reproduce this issue with the version at https://maputnik.github.io/editor/#9.85/51.4922/719.8773, however the link @pathmapper provided https://2155-67726921-gh.circle-artifacts.com/0/artifacts/build/index.html#9.85/51.4922/719.8773 seems to function correctly, can you please confirm?

**Comment by kateler** [comment]
Yes, I confirm it's not happening with @pathmapper 's link.

**Comment by pathmapper** [comment]
Thanks for confirming @kateler, so we could close this issue here?

**Comment by kateler** [comment]
Let me get and build the latest so I can confirm it's fixed for me.

**Comment by kateler** [comment]
Unfortunately, it's still happening for my tiles (which I can't share). 

- I built a new/clean version of maputnik from commit 63ed8c1de34741432cc2adbfd9d1138756579186
- Using a Chrome incognito window, I opened Zoomstack Outdoor and observed the issue was not happening. 
- I loaded my very simple style and observed it is still happening. 

Is there anything you can suggest I dig into, based on the fact it was fixed for the Open Zoomstack tiles between release 1.6.1 and the latest commit?

**Comment by kateler** [comment]
For my tiles (and previously Zoomstack tiles), the feature key is not being properly set by FeaturePropertyPopup. It is being set as '[layername]-undefined' (e.g., 'transportation-undefined'), which then it doesn't like because there are duplicates.

**Comment by orangemug** [comment]
@kateler lets try and dig in, so if you're seeing

> Warning: Encountered two children with the same key

Although an issue, it appears to cause performance problems and haven't actually seen this cause functional issues in the time I've been using React

It looks like you're building a copy of Maputnik yourself, are you running `npm install` before building to grab the latest dependencies, maybe share your build process in the ticket to make sure nothing is going wrong there.

Maybe to reduce the possible differences between our test cases could you load the build https://2155-67726921-gh.circle-artifacts.com/0/artifacts/build/index.html#9.85/51.4922/719.8773 and `right click + save as` then we know we're on the same version. You can run that via `http://localhost:8009/` (instead of `file:///....`) by `cd /path/to/where/you/saved/page` and `python -m SimpleHTTPServer 8009`

Can you also share any errors you are getting from the UI/web inspector, and also anything that might be 'special' about your tiles. Also what source types are you trying to load, is it just vector tiles, or a mix of GeoJSON / vector tiles.

Basically anything that might give us a clue about what differs / causes the issue

**Comment by pathmapper** [comment]
Since `v1.6.1` the Zoomstack source was updated to `v2` in https://github.com/maputnik/editor/commit/58bdd39f9e192573b41171344b2326943f46814d.

Using the previous version with Maputnik `master` the issue is reproducible:

https://2155-67726921-gh.circle-artifacts.com/0/artifacts/build/index.html?style=https://s3-eu-west-1.amazonaws.com/tiles.os.uk/styles/open-zoomstack-outdoor/style.json

So I think we can skip investigating @kateler's build process because it seems that it's related to the tiles.

**Comment by orangemug** [comment]
Thanks @pathmapper, so it turns out

> Although an issue, it appears to cause performance problems and haven't actually seen this cause functional issues

So yeah... this is totally causing the issue, working on a fix now

**Comment by orangemug** [comment]
@kateler hopefully the fix in #610 solves your issue (please test and let us know), you also might want to add IDs to the vector tile features (if they are yours) as it seems to be the norm and might causes issues in other tools to plan to use. There hopefully shouldn't be an issue using Maputnik from now on though 🤞

**Comment by kateler** [comment]
Yes! That solved the problem. Thanks for digging into it. 

For what it's worth, I still get a few errors in my browser console (a lot fewer than before), but the inspect pop-up is functional again. 

The errors are "Uncaught TypeError: Cannot read property 'get' of undefined".

**Comment by orangemug** [comment]
Full error for reference

> Uncaught TypeError: Cannot read property 'get' of undefined
>     at e.queryIntersectsFeature (mapbox-gl.js:formatted:1)
>     at mapbox-gl.js:formatted:1
>     at Hl../node_modules/mapbox-gl/dist/mapbox-gl.js.Hl.loadMatchingFeature (mapbox-gl.js:formatted:1)
>     at x (mapbox-gl.js:formatted:1)
>     at Hl../node_modules/mapbox-gl/dist/mapbox-gl.js.Hl.query (mapbox-gl.js:formatted:1)
>     at Wl../node_modules/mapbox-gl/dist/mapbox-gl.js.Wl.queryRenderedFeatures (mapbox-gl.js:formatted:1)
>     at O (mapbox-gl.js:formatted:1)
>     at i.queryRenderedFeatures (mapbox-gl.js:formatted:1)
>     at r.queryRenderedFeatures (mapbox-gl.js:formatted:1)
>     at MapboxInspect../node_modules/mapbox-gl-inspect/lib/MapboxInspect.js.MapboxInspect._onMousemove (mapbox-gl.js:formatted:1)

The error looks to be in https://github.com/mapbox/mapbox-gl-js/blob/5edbaf1a8186dfa0aad8c955ee958c09492ec855/src/style/style_layer/fill_style_layer.js#L56-L57 when mousing over an area without features in inspect mode.
